simperby/src/node/state_storage.rs

```diff
@@ -0,0 +1,71 @@
+#![allow(unused_mut)]
+#![allow(dead_code)]
+#![allow(unused_variables)]
+
+use crate::*;
+use thiserror::Error;
+
+#[derive(Error, Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub enum StateStorageError {
+    #[error("invalid operation: {0}")]
+    InvalidOperation(String),
+    #[error("storage error: {0}")]
+    StorageError(simperby_kv_storage::Error),
+    /// When an essential item is missing or corrupted.
+    #[error("malformed storage: {0}")]
+    StorageIntegrityError(String),
+}
+
+type Error = StateStorageError;
+
+pub(crate) struct StateStorage {
+    pub storage: Box<dyn KVStorage>,
+}
+
+impl StateStorage {
+    /// Assuming an empty storage, creates the genesis state.
+    pub async fn genesis(
+        mut storage: Box<dyn KVStorage>,
+        genesis_info: GenesisInfo,
+    ) -> Result<Self, Error> {
+        unimplemented!()
+    }
+
+    /// Assuming an existing state, check the basic validty of the state.
+    pub async fn new(storage: Box<dyn KVStorage>) -> Result<Self, Error> {
+        unimplemented!()
+    }
+
+    pub async fn delegation_state_hash(&self) -> Result<Hash256, Error> {
+        unimplemented!()
+    }
+
+    pub async fn state_root(&self) -> Result<Hash256, Error> {
+        unimplemented!()
+    }
+
+    /// Atomically inserts data, hash of it, and updates the state root.
+    pub async fn insert_record(&mut self, key: Hash256, data: &[u8]) -> Result<(), Error> {
```

Comment:
Why don't we hash the data inside this function to get the key?

Reply:
Maybe because the key is not the hash of the data?
(`StateTransition::InsertData { key: String, value: Vec<u8> }`)